[infinitest-intellij] NullPointerException in applyInformationWithProgress() in Idea 2017.2

Occasionally I observe the following exception using Infinitest plugin 5.1.116 with Intelluj Idea Community 2017.2 (build 172.3317.53). In general the plugin seems to work fine.

```
Caused by: java.lang.NullPointerException
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass
             .applyInformationWithProgress(GeneralHighlightingPass.java:272)
```

I haven't seen that in the previous version, but I was using older patched version due to #192, so it might not be affected.

The full stacktrace:

```
java.lang.reflect.InvocationTargetException
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.infinitest.EventNormalizer$1$1.run(EventNormalizer.java:150)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:756)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:726)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:821)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:649)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:365)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.infinitest.EventNormalizer$1$1.run(EventNormalizer.java:144)
	... 17 more
Caused by: java.lang.NullPointerException
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.applyInformationWithProgress(GeneralHighlightingPass.java:272)
	at com.intellij.codeInsight.daemon.impl.ProgressableTextEditorHighlightingPass.doApplyInformationToEditor(ProgressableTextEditorHighlightingPass.java:97)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.applyInformationToEditor(TextEditorHighlightingPass.java:115)
	at org.infinitest.intellij.plugin.launcher.FileEditorListener.testRunCompleted(FileEditorListener.java:44)
	at org.infinitest.intellij.plugin.launcher.InfinitestPresenter.onComplete(InfinitestPresenter.java:191)
	at org.infinitest.intellij.plugin.launcher.InfinitestPresenter.testRunComplete(InfinitestPresenter.java:162)
	... 22 more
```

This is still happening with updated builds.  I can't pin point the exact trigger that causes it to throw this error, but it appears to be happening while watching for changes.  Doesn't appear to block running of the tests though.

## Setup Details:
**IDE:** IntelliJ IDEA 2017.3.4 (Ultimate Edition) - Build #IU-173.4548.28, built on January 29, 2018
**JVM Version:** JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o - JRE: 1.8.0_152-release-1024-b11 x86_64
**Infinitest Version:** 5.0.1.116
**Operating System:** macOS High Sierra 10.13.3 

### Exception Stacktrace:
```
java.lang.reflect.InvocationTargetException
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.infinitest.EventNormalizer$1$1.run(EventNormalizer.java:150)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:762)
	at java.awt.EventQueue.access$500(EventQueue.java:98)
	at java.awt.EventQueue$3.run(EventQueue.java:715)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:732)
	at com.intellij.ide.IdeEventQueue.j(IdeEventQueue.java:822)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:650)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:366)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.GeneratedMethodAccessor594.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.infinitest.EventNormalizer$1$1.run(EventNormalizer.java:144)
	... 17 more
Caused by: java.lang.NullPointerException
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.applyInformationWithProgress(GeneralHighlightingPass.java:274)
	at com.intellij.codeInsight.daemon.impl.ProgressableTextEditorHighlightingPass.doApplyInformationToEditor(ProgressableTextEditorHighlightingPass.java:97)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.applyInformationToEditor(TextEditorHighlightingPass.java:115)
	at org.infinitest.intellij.plugin.launcher.FileEditorListener.testRunCompleted(FileEditorListener.java:44)
	at org.infinitest.intellij.plugin.launcher.InfinitestPresenter.onComplete(InfinitestPresenter.java:191)
	at org.infinitest.intellij.plugin.launcher.InfinitestPresenter.testRunComplete(InfinitestPresenter.java:162)
	... 21 more
```

@szpak it also happens if you have only one IDE open:
```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.infinitest.EventNormalizer$1$1.run(EventNormalizer.java:150)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:764)
	at java.awt.EventQueue.access$500(EventQueue.java:98)
	at java.awt.EventQueue$3.run(EventQueue.java:715)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:734)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:741)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:690)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:385)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.infinitest.EventNormalizer$1$1.run(EventNormalizer.java:144)
	... 17 more
Caused by: java.lang.NullPointerException
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.applyInformationWithProgress(GeneralHighlightingPass.java:256)
	at com.intellij.codeInsight.daemon.impl.ProgressableTextEditorHighlightingPass.doApplyInformationToEditor(ProgressableTextEditorHighlightingPass.java:98)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.applyInformationToEditor(TextEditorHighlightingPass.java:114)
	at org.infinitest.intellij.plugin.launcher.FileEditorListener.testRunCompleted(FileEditorListener.java:44)
	at org.infinitest.intellij.plugin.launcher.InfinitestPresenter.onComplete(InfinitestPresenter.java:191)
	at org.infinitest.intellij.plugin.launcher.InfinitestPresenter.testRunComplete(InfinitestPresenter.java:162)
	... 22 more
```

I got a similar error with IDEA 2022.2 (only one window open)
They seem to have updated the error message and now it says what might be the actual problem

```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.infinitest.EventNormalizer$1$1.run(EventNormalizer.java:150)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:779)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:730)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:724)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:749)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:898)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:746)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:439)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:791)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:438)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:105)
	at com.intellij.ide.IdeEventQueue.performActivity(IdeEventQueue.java:604)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:436)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:881)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:484)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.infinitest.EventNormalizer$1$1.run(EventNormalizer.java:144)
	... 23 more
Caused by: java.lang.IllegalStateException: Must not instantiate passes in EDT
	at com.intellij.codeInsight.daemon.impl.TextEditorHighlightingPassRegistrarImpl.instantiatePasses(TextEditorHighlightingPassRegistrarImpl.java:168)
	at com.intellij.codeInsight.daemon.impl.TextEditorBackgroundHighlighter.getPasses(TextEditorBackgroundHighlighter.java:75)
	at com.intellij.codeInsight.daemon.impl.TextEditorBackgroundHighlighter.createPassesForEditor(TextEditorBackgroundHighlighter.java:80)
	at com.intellij.codeInsight.daemon.impl.TextEditorBackgroundHighlighter.createPassesForEditor(TextEditorBackgroundHighlighter.java:24)
	at org.infinitest.intellij.plugin.launcher.FileEditorListener.testRunCompleted(FileEditorListener.java:43)
	at org.infinitest.intellij.plugin.launcher.InfinitestPresenter.onComplete(InfinitestPresenter.java:191)
	at org.infinitest.intellij.plugin.launcher.InfinitestPresenter.testRunComplete(InfinitestPresenter.java:162)
	... 28 more

```

--- a/infinitest-intellij/src/main/java/org/infinitest/intellij/idea/language/InfinitestHighlightingPassFactory.java
+++ b/infinitest-intellij/src/main/java/org/infinitest/intellij/idea/language/InfinitestHighlightingPassFactory.java
@@ -33,9 +33,14 @@ import com.intellij.codeHighlighting.*;
 import com.intellij.openapi.editor.*;
 import com.intellij.openapi.module.*;
 import com.intellij.openapi.module.Module;
+import com.intellij.openapi.util.Key;
 import com.intellij.psi.*;
 
 public class InfinitestHighlightingPassFactory implements TextEditorHighlightingPassFactory {
+	/**
+	 * The key to retrieve the {@link InfinitestLineMarkersPass} in the {@link Editor}
+	 */
+	public static final Key<InfinitestLineMarkersPass> KEY = new Key<>("InfinitestLineMarkersPass");
 
 	public InfinitestHighlightingPassFactory(TextEditorHighlightingPassRegistrar passRegistrar) {
 		passRegistrar.registerTextEditorHighlightingPass(this, TextEditorHighlightingPassRegistrar.Anchor.LAST, Pass.UPDATE_ALL, true, true);
@@ -48,6 +53,9 @@ public class InfinitestHighlightingPassFactory implements TextEditorHighlighting
 			return null;
 		}
 
-		return new InfinitestLineMarkersPass(module.getProject(), editor.getDocument(), editor.getMarkupModel());
+		InfinitestLineMarkersPass pass = new InfinitestLineMarkersPass(module.getProject(), editor.getDocument(), editor.getMarkupModel());
+		editor.putUserData(KEY, pass);
+		
+		return pass;
 	}
 }

--- a/infinitest-intellij/src/main/java/org/infinitest/intellij/plugin/launcher/FileEditorListener.java
+++ b/infinitest-intellij/src/main/java/org/infinitest/intellij/plugin/launcher/FileEditorListener.java
@@ -27,8 +27,13 @@
  */
 package org.infinitest.intellij.plugin.launcher;
 
-import com.intellij.codeHighlighting.*;
-import com.intellij.openapi.fileEditor.*;
+import org.infinitest.intellij.idea.language.InfinitestHighlightingPassFactory;
+import org.infinitest.intellij.idea.language.InfinitestLineMarkersPass;
+
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.fileEditor.FileEditor;
+import com.intellij.openapi.fileEditor.FileEditorManager;
+import com.intellij.openapi.fileEditor.TextEditor;
 
 public class FileEditorListener implements PresenterListener {
 	FileEditorManager fileEditorManager;
@@ -40,8 +45,15 @@ public class FileEditorListener implements PresenterListener {
 	@Override
 	public void testRunCompleted() {
 		for (FileEditor fileEditor : fileEditorManager.getSelectedEditors()) {
-			for (HighlightingPass highlightingPass : fileEditor.getBackgroundHighlighter().createPassesForEditor()) {
-				highlightingPass.applyInformationToEditor();
+			if (fileEditor instanceof TextEditor) {
+				TextEditor textEditor = (TextEditor) fileEditor;
+				Editor editor = textEditor.getEditor();
+				
+				InfinitestLineMarkersPass pass = editor.getUserData(InfinitestHighlightingPassFactory.KEY);
+
+				if (pass != null) {
+					pass.applyInformationToEditor();
+				}
 			}
 		}
 	}

